pil-analyzer/tests/types.rs

```diff
@@ -561,6 +561,136 @@ fn empty_conditional() {
     type_check(input, &[]);
 }
 
+#[test]
+fn defined_trait() {
+    let input = "
+    trait Add<T> {
+        add: T, T -> T,
+    }
+    impl Add<int> {
+        add: |a, b| a + b,
+    }
+    let r: int = Add::add(3, 4);
+    ";
+    type_check(input, &[("r", "", "int")]);
+}
+
+#[test]
+#[should_panic = "Trait Add not found"]
+fn undefined_trait() {
+    let input = "
+    impl Add<int> {
+        add: |a, b| a + b,
+    }
+    ";
+    type_check(input, &[]);
+}
+
+#[test]
+#[should_panic = "Impls for Add: Types (int, fe) and (int, fe) overlap"]
+fn duplicated_trait() {
+    let input = "
+    trait Add<T, Q> {
+        add: T, T -> Q,
+    }
+    impl Add<int, fe> {
+        add: |a, b| a + b,
```

Comment:
If we ever swap type checking and overlap checking around, this test will not test the overlapping any more. Can you change this so that everything is correct except the overlapping?